pkg/client/dryrun_test.go

```diff
@@ -0,0 +1,226 @@
+		Expect(getClient().Delete(ctx, dep)).NotTo(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
+
+	It("should not delete objects with opts", func() {
+		opts := &client.DeleteOptions{DryRun: []string{"Bye", "Pippa"}}
+
+		Expect(getClient().Delete(ctx, dep, opts)).NotTo(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
+
+	It("should not delete objects via deleteAllOf", func() {
+		opts := []client.DeleteAllOfOption{client.InNamespace(ns), client.MatchingLabels(dep.Labels)}
+
+		Expect(getClient().DeleteAllOf(ctx, dep, opts...)).NotTo(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
+
+	It("should not change objects via update status", func() {
+		changedDep := dep.DeepCopy()
+		changedDep.Status.Replicas = 99
+
+		Expect(getClient().Status().Update(ctx, changedDep)).NotTo(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
+
+	It("should not change objects via update status with opts", func() {
+		changedDep := dep.DeepCopy()
+		changedDep.Status.Replicas = 99
+		opts := &client.UpdateOptions{DryRun: []string{"Bye", "Pippa"}}
+
+		Expect(getClient().Status().Update(ctx, changedDep, opts)).NotTo(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
+
+	It("should not change objects via status patch", func() {
+		changedDep := dep.DeepCopy()
+		changedDep.Status.Replicas = 99
+
+		Expect(getClient().Status().Patch(ctx, changedDep, client.MergeFrom(dep))).ToNot(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
+
+	It("should not change objects via status patch with opts", func() {
+		changedDep := dep.DeepCopy()
+		changedDep.Status.Replicas = 99
+
+		opts := &client.PatchOptions{DryRun: []string{"Bye", "Pippa"}}
+
+		Expect(getClient().Status().Patch(ctx, changedDep, client.MergeFrom(dep), opts)).ToNot(HaveOccurred())
+
+		actual, err := clientset.AppsV1().Deployments(ns).Get(dep.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		Expect(actual).NotTo(BeNil())
+		Expect(actual).To(BeEquivalentTo(dep))
+	})
```

Comment:
Reading https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/ I see that one of the advantages of supporting a dry-run, besides not being persisted, is

> the request is still processed as typical request: the fields are defaulted, the object is validated, it goes through the validation admission chain, and through the mutating admission chain, and then the final object is returned to the user as it normally would, without being persisted.

I think a test case where a create or an update failed validation would be a good addition.